arangod/Aql/OptimizerRulesCluster.cpp

```diff
@@ -0,0 +1,464 @@
+  if(oldNode == plan->root()) {
+    for(auto* dep : oldNode->getDependencies()) {
+      newNode->addDependency(dep);
+    }
+    plan->root(newNode,true);
+  } else {
+    // replaceNode does not seem to work well with subqueries
+    // if the subqueries root is replaced.
+    // It looks like Subquery node will still point to
+    // the old node.
+
+    TRI_ASSERT(oldNode != plan->root());
+    plan->replaceNode(oldNode, newNode);
+    TRI_ASSERT(!oldNode->hasDependency());
+    TRI_ASSERT(!oldNode->hasParent());
+  }
+}
+
+bool substituteClusterSingleDocumentOperationsIndex(Optimizer* opt,
+                                                    ExecutionPlan* plan,
+                                                    OptimizerRule const* rule) {
+  bool modified = false;
+  SmallVector<ExecutionNode*>::allocator_type::arena_type a;
+  SmallVector<ExecutionNode*> nodes{a};
+  plan->findNodesOfType(nodes, EN::INDEX, false);
+
+  if(nodes.size() != 1){
+    return modified;
+  }
+
+  for(auto* node : nodes){
+
+    if(!depIsSingletonOrConstCalc(node)){
+      continue;
+    }
+
+    Index* index = hasSingleIndexHandle(node, Index::TRI_IDX_TYPE_PRIMARY_INDEX);
+    if (index){
+      IndexNode* indexNode = static_cast<IndexNode*>(node);
+      auto binaryCompares = hasBinaryCompare(node);
+      std::string key = getFirstKey(binaryCompares);
+      if(key.empty()){
+        continue;
+      }
+
+      auto* parentModification = hasSingleParent(node,{EN::INSERT, EN::REMOVE, EN::UPDATE, EN::REPLACE});
+
+      if (parentModification){
+        auto mod = static_cast<ModificationNode*>(parentModification);
+        auto parentType = parentModification->getType();
+        auto const& vec = mod->getVariablesUsedHere();
+
+        Variable const* update = nullptr;
+        Variable const* keyVar = nullptr;
+
+        if ( parentType == EN::REMOVE) {
+          keyVar = vec.front();
+          TRI_ASSERT(vec.size() == 1);
+        } else {
+          update = vec.front();
+          if(vec.size() > 1){
+            keyVar = vec.back();
+          }
+        }
+
+        if(keyVar && indexNode->outVariable()->id != keyVar->id){
+          continue;
+        }
+
+        if(!keyVar){
+          if (update && indexNode->outVariable()->id == update->id){
+            // the update document is already described by the key provided
+            // in the index condition
+            update = nullptr;
+          } else {
+            continue;
+          }
+        }
+
+        ExecutionNode* singleOperationNode = plan->registerNode(
+          new SingleRemoteOperationNode(
+            plan, plan->nextId(),
+            parentType,
+            true, key, mod->collection(),
+            mod->getOptions(),
+            update,
+            nullptr,
+            mod->getOutVariableOld(),
+            mod->getOutVariableNew()
+          )
+        );
+
+        if(!parentIsReturnOrConstCalc(mod)){
+          continue;
```

Comment:
if we hit this `continue` here, what will happen with the `singleOperationNode` that we registered in the plan already?

it will not be linked anywhere in the plan. shouldn't the node creation be moved below this `if`?